iroh/src/magicsock/transports/relay.rs

```diff
@@ -0,0 +1,472 @@
+
+        let my_node_id = config.secret_key.public();
+        let my_relay = config.my_relay.clone();
+
+        let relay_actor = RelayActor::new(config, relay_datagram_recv_queue.clone());
+
+        let actor_handle = AbortOnDropHandle::new(task::spawn(
+            async move {
+                relay_actor
+                    .run(actor_receiver, relay_datagram_send_rx)
+                    .await;
+            }
+            .instrument(info_span!("relay-actor")),
+        ));
+
+        Self {
+            relay_datagram_recv_queue,
+            relay_datagram_send_channel: relay_datagram_send_tx,
+            actor_sender,
+            _actor_handle: actor_handle,
+            my_relay,
+            my_node_id,
+        }
+    }
+
+    fn send_relay_actor(&self, msg: RelayActorMessage) {
+        match self.actor_sender.try_send(msg) {
+            Ok(_) => {}
+            Err(mpsc::error::TrySendError::Closed(_)) => {
+                warn!("unable to send to relay actor, already closed");
+            }
+            Err(mpsc::error::TrySendError::Full(_)) => {
+                warn!("dropping message for relay actor, channel is full");
+            }
+        }
+    }
+
+    pub(super) fn create_io_poller(&self) -> RelayDatagramSendChannelSender {
+        self.relay_datagram_send_channel.clone()
+    }
+
+    pub(super) fn poll_send(
+        &self,
+        dest_url: RelayUrl,
+        dest_node: NodeId,
+        transmit: &Transmit<'_>,
+    ) -> Poll<io::Result<()>> {
+        let contents = split_packets(transmit);
+
+        let msg = RelaySendItem {
+            remote_node: dest_node,
+            url: dest_url.clone(),
+            datagrams: contents,
+        };
+
+        match self.relay_datagram_send_channel.try_send(msg) {
+            Ok(_) => {
+                trace!(node = %dest_node.fmt_short(), relay_url = %dest_url,
+                       "send relay: message queued");
+                Poll::Ready(Ok(()))
+            }
+            Err(mpsc::error::TrySendError::Closed(_)) => {
+                error!(node = %dest_node.fmt_short(), relay_url = %dest_url,
+                      "send relay: message dropped, channel to actor is closed");
+                Poll::Ready(Err(io::Error::new(
+                    io::ErrorKind::ConnectionReset,
+                    "channel to actor is closed",
+                )))
+            }
+            Err(mpsc::error::TrySendError::Full(_)) => {
+                warn!(node = %dest_node.fmt_short(), relay_url = %dest_url,
+                      "send relay: message dropped, channel to actor is full");
+                Poll::Pending
+            }
+        }
+    }
+
+    pub(super) fn poll_recv(
+        &self,
+        cx: &mut Context,
+        bufs: &mut [io::IoSliceMut<'_>],
+        metas: &mut [quinn_udp::RecvMeta],
+        source_addrs: &mut [Addr],
+    ) -> Poll<io::Result<usize>> {
+        let mut num_msgs = 0;
+        'outer: for ((buf_out, meta_out), addr) in bufs
```

Comment:
Not sure why this has a label?  It's not a nested loop?  I guess it was before.

Reply:
fixed